### PR TITLE
Configure languageVersion via toolchain

### PR DESCRIPTION
**Pull request by Goooler**
Follow up #5614.

https://kotlinlang.org/docs/gradle-configure-project.html#gradle-java-toolchains-support
https://jakewharton.com/build-on-latest-java-test-through-lowest-java

### CODE REVIEW

**Review by schalkms — APPROVED** [review]

**Review by 3flex — CHANGES_REQUESTED** [review]
I did something very similar in https://github.com/detekt/detekt/pull/5676 and reverted in https://github.com/detekt/detekt/pull/5685 because:

> The original PR only was only half a solution - it meant that all builds effectively ran on Java 1.8 regardless of which Java version the Gradle build was running. This is fine for the main source build but not for tests, which we want to run on multiple Java versions.

If you can make updates to allow for that then I'd be happy to approve but there'll likely be config changes to a few places in the build and potentially the GitHub workflows as well.

**Comment by Goooler** [comment]
The solution might be [Build on latest Java, test through lowest Java](https://jakewharton.com/build-on-latest-java-test-through-lowest-java), but there is a `NoSuchMethodError` in `detekt-compiler-plugin` test

```sh
CompilerTest > with a source file that contains violations() FAILED
    java.lang.NoSuchMethodError: 'void org.jetbrains.kotlin.cli.common.arguments.K2JVMCompilerArguments.setSingleModule(boolean)'
        at app//com.tschuchort.compiletesting.KotlinCompilation$commonK2JVMArgs$1.invoke(KotlinCompilation.kt:359)
        at app//com.tschuchort.compiletesting.KotlinCompilation$commonK2JVMArgs$1.invoke(KotlinCompilation.kt:300)
        at app//com.tschuchort.compiletesting.AbstractKotlinCompilation.commonArguments(AbstractKotlinCompilation.kt:133)
        at app//com.tschuchort.compiletesting.KotlinCompilation.commonK2JVMArgs(KotlinCompilation.kt:300)
        at app//com.tschuchort.compiletesting.KotlinCompilation.compileJvmKotlin(KotlinCompilation.kt:488)
        at app//com.tschuchort.compiletesting.KotlinCompilation.compile(KotlinCompilation.kt:664)
        at app//io.github.detekt.compiler.plugin.util.CompilerTestUtils.compile(CompilerTestUtils.kt:21)
        at app//io.github.detekt.compiler.plugin.CompilerTest.with a source file that contains violations(CompilerTest.kt:11)
```

**Comment by 3flex** [comment]
The `test` task is currently disabled on `detekt-compiler-plugin`, so these auto generated tasks should be disabled on that project too.

**Comment by Goooler** [comment]
Disabled. Let's see the impacts on build time.

**Comment by 3flex** [comment]
Yeah that was going to be my next point, but I wanted to see the actual CI runtimes first.

They're not good (which is expected). The loss of parallelism for test execution on the different JVM versions really hurts. This tradeoff isn't worth it just to use toolchains.

**Comment by cortinico** [comment]
> The loss of parallelism for test execution on the different JVM versions really hurts.

+1 for this. I'd rather keep explicit Java versions in the CI matrix as it's also easier for users to understand which version of Java broke

**Comment by Goooler** [comment]
Agreed, I'll leave this PR open until we find new way to optimize.

**Comment by Goooler** [comment]
I'm not sure why builds on Windows are still failing after applying `foojay-resolver-convention`, but the build time on Ubuntu looks resonable.

**Comment by Goooler** [comment]
It looks fine for now.

File: detekt-gradle-plugin/settings.gradle.kts [hunk 1]
```diff
@@ -14,4 +14,5 @@ dependencyResolutionManagement {
 
 plugins {
     id("com.gradle.enterprise") version "3.12.4"
+    id("org.gradle.toolchains.foojay-resolver-convention") version "0.4.0"
```
Comment (cortinico) [inline detekt-gradle-plugin/settings.gradle.kts:17]: Why is this needed?

Comment (Goooler) [reply detekt-gradle-plugin/settings.gradle.kts:17]: For Windows, check https://github.com/detekt/detekt/actions/runs/4464478367/jobs/7840655669

**Comment by chao2zhang** [comment]
Is this PR still blocked? Could we make a call on whether to move forward or to close this?

**Comment by cortinico** [comment]
My 2 c here: I'd love us to use `jvmToolchain` here but I'm not a big fan on adding yet another Gralde plugin to handle the Win case. So I'd rather close this for now and get back to it?

File: detekt-gradle-plugin/settings.gradle.kts (hunk 1 above)
Comment (3flex) [reply detekt-gradle-plugin/settings.gradle.kts:17]: It's not needed for Windows specifically, it's required any time a requested JDK is not found locally. It's a requirement in most cases if using toolchains.

More here: https://docs.gradle.org/current/userguide/toolchains.html#sub:download_repositories

Comment (Goooler) [reply detekt-gradle-plugin/settings.gradle.kts:17]: Yep.

**Comment by 3flex** [comment]
I'm going to close this, though not because of the foojay plugin (that's essentially a requirement to use toolchains outside of controlled environments where JDKs are all pre installed).

Instead I'd like to wait until this is ready later this year: https://github.com/gradle/build-tool-roadmap/issues/61

We can then just use a standard config to set up the test suites instead of developing a custom one. Things are working ok as they are so there's no real reason to switch yet.

**Pull request closed without merging**